src/main/java/com/comphenix/protocol/wrappers/Either.java

```diff
@@ -0,0 +1,62 @@
+package com.comphenix.protocol.wrappers;
+
+import java.util.Optional;
+import java.util.function.Function;
+
+public abstract class Either<L, R> {
+
+    public static class Left<L, R> extends Either<L, R> {
```

Comment:
is there a reason for having them be separate classes vs a single class? i'd imagine you could get the same functionality in one class with an exclusive or

Reply:
could have static factory methods `left` and `right`

Reply:
>could have static factory methods left and right

I'll change that.

>is there a reason for having them be separate classes vs a single class? i'd imagine you could get the same functionality in one class with an exclusive or

The implementation is similar to Mojang's implementation. Having separate classes allows us to have a single field while also being type-safe. Left and Right could be/are different types. Therefore we either use generic Object or two fields, where one them is null.

Pseudo:
```java
class Either {

private final Object value;
private boolean left;

public Optional<L> left() {
    return left ? Optional.empty() : Optional.of((L) value);
}

public Optional<R> right() {
    return left ? Optional.of((R) value) : Optional.empty();
}
```

or

```java
class Either {

private final L left;
private final R right;

public Optional<L> left() {
    return Optional.ofNullable(left);
}

public Optional<R> right() {
    return Optional.ofNullable(right);
}
```
Note: The constructor could then provide the exclusivity.

Reply:
I'm good with leaving them as separate classes if that's what Mojang does. Might still be worth having those factory methods in Either